geoh5py/objects/surface.py

```diff
@@ -18,61 +18,74 @@
 from __future__ import annotations
 
 import uuid
+import warnings
 
 import numpy as np
 
 from .cell_object import CellObject
-from .object_base import ObjectType
 
 
 class Surface(CellObject):
     """
     Surface object defined by vertices and cells
     """
 
-    __TYPE_UID = uuid.UUID(
+    _TYPE_UID = uuid.UUID(
         fields=(0xF26FEBA3, 0xADED, 0x494B, 0xB9, 0xE9, 0xB2BBCBE298E1)
     )
 
-    def __init__(self, object_type: ObjectType, **kwargs):
-        self._cells: np.ndarray | None = None
+    def validate_cells(self, indices: list | tuple | np.ndarray | None) -> np.ndarray:
+        """
+        Validate or generate cells made up of triplets of vertices making
+            up triangles.
 
-        super().__init__(object_type, **kwargs)
+        :param indices: Array of indices, shape(*, 3). If None provided, the
+            vertices are connected sequentially.
 
-    @property
-    def cells(self) -> np.ndarray | None:
-        """
-        Array of vertices index forming triangles
-        :return cells: :obj:`numpy.array` of :obj:`int`, shape ("*", 3)
+        :return: Array of indices defining connecting vertices.
         """
-        if getattr(self, "_cells", None) is None:
-            if self.on_file:
-                self._cells = self.workspace.fetch_array_attribute(self)
-
-        return self._cells
-
-    @cells.setter
-    def cells(self, indices: list | np.ndarray | None):
-        if isinstance(indices, list):
-            indices = np.vstack(indices)
-
-        if self._cells is not None and (
-            indices is None or indices.shape[0] < self._cells.shape[0]
-        ):
-            raise ValueError(
-                "Attempting to assign 'cells' with fewer values. "
-                "Use the `remove_cells` method instead."
+        if isinstance(indices, (tuple | list)):
+            indices = np.array(indices, ndmin=2)
+
+        if indices is None:
+            n_vert = self.vertices.shape[0]
+            indices = np.c_[
+                np.arange(0, n_vert - 2), np.arange(1, n_vert - 1), np.arange(2, n_vert)
+            ].astype("uint32")
+
+        if not isinstance(indices, np.ndarray):
+            raise AttributeError(
+                "Attribute 'cells' must be provided as type numpy.ndarray, list or tuple."
             )
 
-        if indices.ndim != 2 or indices.shape[1] != 3:
-            raise ValueError("Array of cells should be of shape (*, 3).")
+        if indices.ndim != 2 or indices.shape[-1] != 3:
+            raise ValueError("Array of 'cells' should be of shape (*, 3).")
 
         if not np.issubdtype(indices.dtype, np.integer):
             raise TypeError("Indices array must be of integer type")
 
```

Comment:
Same question as curve:

Is there more validation to do? Like a node can only be call 3 times, there can't be orphans nodes etc.

Reply:
Already done on the base class